VERSION-RANGE-SPEC.rst

```diff
@@ -0,0 +1,989 @@
+
+- **Versions are unique**. Each ``version`` must be unique in a range and can
+  occur only once in any ``<version-constraint>`` of a range specifier,
+  irrespective of its comparators. Tools must report an error for duplicated
+  versions.
+
+- **There is only one star**: "*" must only occur once and alone in a range,
+  without any other constraint or version.
+
+Starting from a de-duplicated and sorted list of constraints, these extra rules
+apply to the comparators of any two contiguous constraints to be valid:
+
+- "!=" constraint can be followed by a constraint using any comparator, i.e.,
+  any of "=", "!=", ">", ">=", "<", "<=" as comparator (or no constraint).
+
+Ignoring all constraints with "!=" comparators:
+
+- A "=" constraint must be followed only by a constraint with one of "=", ">",
+  ">=" as comparator (or no constraint).
+
+And ignoring all constraints with "=" or "!=" comparators, the sequence of
+constraint comparators must be an alternation of greater and lesser comparators:
+
+- "<" and "<=" must be followed by one of ">", ">=" (or no constraint).
+- ">" and ">=" must be followed by one of "<", "<=" (or no constraint).
+
+Tools must report an error for such invalid ranges.
+
+
+Parsing and validating version range specifiers
+-------------------------------------------------
+
+To parse a version range specifier string:
+
+- Remove all spaces and tabs.
+- Start from left, and split once on colon ":".
+- The left hand side is the URI-scheme that must be lowercase.
+  - Tools must validate that the URI-scheme value is ``vers``.
+- The right hand side is the specifier.
+
+- Split the specifier from left once on a slash "/".
+
+- The left hand side is the <versioning-scheme> that must be lowercase.
+  Tools should validate that the <versioning-scheme> is a known scheme.
+
+- The right hand side is a list of one or more constraints.
+  Tools must validate that this constraints string is not empty ignoring spaces.
+
+- If the constraints string is equal to "*", the ``<version-constraint>`` is "*".
+  Parsing is done and no further processing is needed for this ``vers``. A tool
+  should report an error if there are extra characters beyond "*". 
+
+- Strip leading and trailing pipes "|" from the constraints string.
+- Split the constraints on pipe "|". The result is a list of ``<version-constraint>``.
+  Consecutive pipes must be treated as one and leading and trailing pipes ignored.
+
+- For each ``<version-constraint>``:
+  - Determine if the ``<version-constraint>`` starts with one of the two comparators:
+
+    - If it starts with ">=", then the comparator is ">=".
+    - If it starts with "<=", then the comparator is "<=".
+    - If it starts with "!=", then the comparator is "!=".
+    - If it starts with "<",  then the comparator is "<".
+    - If it starts with ">",  then the comparator is ">".
+
+    - Remove the comparator from ``<version-constraint>`` string start. The
+      remaining string is the version.
+
```

Comment:
So `>>1.2.3` will be split into `>` as comparator and `>1.2.3` as version, and `<>1.2.3` into `<` and `>1.2.3`, correct?

IIRC, version strings must start with a digit, so why not splitting a `<version-constraint>` at the first digit which occurs?

Also, please consider translating comparators which are used in other notations by expanding the fixed list by:
- `- If it starts with "<>",  then the comparator is "!=".`
- `- If it starts with ">>",  then the comparator is ">".`
- `- If it starts with "<<",  then the comparator is "<".`

IMO, all other comparators (after splitting a `<version-constraint>` at the first digit) should be specified as invalid and result in an error message.